File: crud-item-list/src/main/java/com/gurunars/crud_item_list/UiThrottleBuffer.java
package com.gurunars.crud_item_list;

import android.os.Handler;

/* Allows to buffer runnable execution to prevent spamming events of the same type. */
class UiThrottleBuffer {

    private static final int TIMEOUT = 500;
    private final Handler handler = new Handler();
    private Runnable currentCallback;

    private void cancel() {
        handler.removeCallbacks(currentCallback);
    }

    void shutdown() {
        cancel();
        if (currentCallback != null) {
            currentCallback.run();
        }
    }

    void call(final Runnable runnable) {
        cancel();
        currentCallback = new Runnable() {
            @Override
            public void run() {
                runnable.run();
                currentCallback = null;
            }
        };
        handler.postDelayed(currentCallback, TIMEOUT);
    }

}

File: crud-item-list/src/main/kotlin/com/gurunars/crud_item_list/UiThrottleBuffer.kt
package com.gurunars.crud_item_list

import android.os.Handler

/* Allows to buffer runnable execution to prevent spamming events of the same type. */
internal class UiThrottleBuffer {
    private val handler = Handler()
    private var currentCallback = {}

    private fun cancel() {
        handler.removeCallbacks(currentCallback)
    }

    fun shutdown() {
        cancel()
        currentCallback()
    }

    fun call(runnable: () -> Unit) {
        cancel()
        currentCallback = {
            runnable()
            currentCallback = {}
        }
        handler.postDelayed(currentCallback, TIMEOUT.toLong())
    }

    companion object {
        private val TIMEOUT = 500
    }

}
